api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java

```diff
@@ -0,0 +1,201 @@
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.TraceState;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Implementation of the B3 propagation protocol. See <a
+ * href=https://github.com/openzipkin/b3-propagation>openzipkin/b3-propagation</a>.
+ */
+@Immutable
+public class B3Propagator implements HttpTextFormat<SpanContext> {
+  private static final Logger logger = Logger.getLogger(HttpTraceContext.class.getName());
+
+  static final String TRACE_ID_HEADER = "X-B3-TraceId";
+  static final String SPAN_ID_HEADER = "X-B3-SpanId";
+  static final String SAMPLED_HEADER = "X-B3-Sampled";
+  static final String TRUE_INT = "1";
+  static final String FALSE_INT = "0";
+  static final String COMBINED_HEADER = "b3";
+  static final String COMBINED_HEADER_DELIMITER = "-";
+
+  private static final int MAX_TRACE_ID_LENGTH = 2 * TraceId.getSize();
+  private static final int MAX_SPAN_ID_LENGTH = 2 * SpanId.getSize();
+
+  private static final List<String> FIELDS =
+      Collections.unmodifiableList(Arrays.asList(TRACE_ID_HEADER, SPAN_ID_HEADER, SAMPLED_HEADER));
+
+  private final boolean singleHeader;
+
+  /** Creates a new instance of {@link B3Propagator}. Defaults to use multiple headers. */
+  public B3Propagator() {
+    this(false);
+  }
+
+  /**
+   * Creates a new instance of {@link B3Propagator}.
+   *
+   * @param singleHeader whether to use single or multiple headers.
+   */
+  public B3Propagator(Boolean singleHeader) {
+    this.singleHeader = singleHeader;
+  }
+
+  @Override
+  public List<String> fields() {
+    return FIELDS;
+  }
+
+  @Override
+  public <C> void inject(SpanContext spanContext, C carrier, Setter<C> setter) {
+    checkNotNull(spanContext, "spanContext");
+    checkNotNull(setter, "setter");
+    checkNotNull(carrier, "carrier");
+
+    String sampled = spanContext.getTraceFlags().isSampled() ? TRUE_INT : FALSE_INT;
+
+    if (singleHeader) {
+      setter.set(
+          carrier,
+          COMBINED_HEADER,
+          spanContext.getTraceId().toLowerBase16()
+              + COMBINED_HEADER_DELIMITER
+              + spanContext.getSpanId().toLowerBase16()
+              + COMBINED_HEADER_DELIMITER
+              + sampled);
+    } else {
+      setter.set(carrier, TRACE_ID_HEADER, spanContext.getTraceId().toLowerBase16());
+      setter.set(carrier, SPAN_ID_HEADER, spanContext.getSpanId().toLowerBase16());
+      setter.set(carrier, SAMPLED_HEADER, sampled);
+    }
+  }
+
+  @Override
+  @SuppressWarnings("StringSplitter")
+  public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter) {
+    checkNotNull(carrier, "carrier");
+    checkNotNull(getter, "getter");
+
+    String traceId;
+    String spanId;
+    String sampled;
+    if (singleHeader) {
```

Comment:
This should be broken up into at least 2 separate methods. This method is definitely too long. One method for the single header case that returns the context, and one for the multi-header case that returns the context will make this code much easier to read. Thanks!

Reply:
I support this. The TraceContext one is refactored as it was too long and it was (relatively) hard to read.